### PR TITLE
[Merged by Bors] - feat(CategoryTheory/Galois): Fibers are represented by Galois objects

### PR DESCRIPTION
**Pull request by chrisflav**
We show that if `X` is an object in a Galois category, then there exists an object `A` and a point `a` in the fiber of `A` such that `A` is Galois and that the evaluation map at `a` from `A ⟶ X` to the fiber of `X` is bijective.

This is the main input in the pro-representability of fiber functors.

---

- [x] depends on: #10419

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 1]
```diff
@@ -21,6 +21,13 @@ is represented by a Galois object (TODO).
   connected component.
 * `connected_component_unique`: Up to isomorphism, for each element `x` in the fiber of `X` there
   is only one connected component whose fiber contains `x`.
+* `exists_galois_representative`: The fiber of `X` is represented by some Galois object `A`:
+  Evaluation at some `a` in the fiber of `A` induces a bijection `A ⟶ X` to `F.obj X`.
+
+## References
+
+* H. W. Lenstra. Galois theory for schemes
+  <https://websites.math.leidenuniv.nl/algebra/GSchemes.pdf>
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:30]: Do we have an entry for this in `docs/references.bib`? If not, please add it.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 2]
```diff
@@ -160,6 +167,144 @@ lemma connected_component_unique {X A B : C} [IsConnected A] [IsConnected B] (a
...
+a Galois object `A` and an element `a` in the fiber of `A` such that the
+evaluation at `a` from `A ⟶ X` to `F.obj X` is bijective.
+
+To show this we consider the product `∏ (fun _ : F.obj X ↦ X)` and let `A`
+be the connected component whose fiber contains the element `a` in the fiber of the self product
+that has at each index `x : F.obj X` the element `x`.
+
+This `A` is Galois and evaluation at `a` is bijective.
+
+Reference: Lenstra, 3.14
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:184]: And refer to it here please.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 3]
```diff
@@ -160,6 +167,144 @@ lemma connected_component_unique {X A B : C} [IsConnected A] [IsConnected B] (a
...
+  · simp only [selfProdProj, map_comp, FintypeCat.comp_apply]; rfl
+  · dsimp only [selfProdPermIncl, Pi.whiskerEquiv]
+    rw [map_comp, FintypeCat.comp_apply, h]
+    convert_to F.map (selfProdProj u t) b =
+      (F.map (Pi.map' (fiberPerm h b) fun _ ↦ 𝟙 X) ≫
+      F.map (Pi.π (fun _ ↦ X) t)) (mkSelfProdFib F X)
+    rw [← map_comp, Pi.map'_comp_π, Category.comp_id, mkSelfProdFib_map_π F X (fiberPerm h b t)]
+    rfl
+
+/- `fiso` is obtained by considering `u` and `selfProdPermIncl h b`. Both
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:256]: What does `fiso` stand for?

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 4]
```diff
@@ -160,6 +167,144 @@ lemma connected_component_unique {X A B : C} [IsConnected A] [IsConnected B] (a
...
+    rw [hfi1, ← hfi2]
+    exact congr_fun (F.mapIso fi2).hom_inv_id y
+  · refine ⟨evaluationInjective_of_isConnected F A X a, ?_⟩
+    intro x
+    use u ≫ Pi.π _ x
+    exact (selfProdProj_fiber h1) x
+
+/-- Any element in the fiber of an object `X` is the evaluation of a morphism from a
+Galois object. -/
+lemma exists_map_from_galois_of_fiber (X : C) (x : F.obj X) :
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:288]:
```suggestion
lemma exists_hom_from_galois_of_fiber (X : C) (x : F.obj X) :
```

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 5]
```diff
@@ -160,6 +167,144 @@ lemma connected_component_unique {X A B : C} [IsConnected A] [IsConnected B] (a
...
+    ∃ (A : C) (f : A ⟶ X) (a : F.obj A), IsGalois A ∧ F.map f a = x := by
+  obtain ⟨A, a, h1, h2⟩ := exists_galois_representative F X
+  obtain ⟨f, hf⟩ := h2.surjective x
+  exact ⟨A, f, a, h1, hf⟩
+
+/-- Any object with non-empty fiber admits a map from a Galois object. -/
+lemma exists_map_from_galois_of_fiber_nonempty (X : C) (h : Nonempty (F.obj X)) :
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:295]:
```suggestion
lemma exists_hom_from_galois_of_fiber_nonempty (X : C) (h : Nonempty (F.obj X)) :
```

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 6]
```diff
@@ -160,6 +167,144 @@ lemma connected_component_unique {X A B : C} [IsConnected A] [IsConnected B] (a
...
+    ∃ (A : C) (_ : A ⟶ X), IsGalois A := by
+  obtain ⟨x⟩ := h
+  obtain ⟨A, f, a, h1, _⟩ := exists_map_from_galois_of_fiber F X x
+  exact ⟨A, f, h1⟩
+
+/-- Any connected object admits a map from a Galois object. -/
+lemma exists_map_from_galois_of_connected (X : C) [IsConnected X] :
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:302]:
```suggestion
lemma exists_hom_from_galois_of_connected (X : C) [IsConnected X] :
```

File: Mathlib/CategoryTheory/Galois/Decomposition.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean]: Please check if this is what you meant, I am not sure I interpreted the instructions in `docs/references.bib` correctly.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/CategoryTheory/Galois/Decomposition.lean]: Nothing :D I changed the docstring to make it clearer.

File: Mathlib/CategoryTheory/Galois/Decomposition.lean [hunk 7]
```diff
@@ -21,6 +21,13 @@ is represented by a Galois object (TODO).
   connected component.
 * `connected_component_unique`: Up to isomorphism, for each element `x` in the fiber of `X` there
   is only one connected component whose fiber contains `x`.
+* `exists_galois_representative`: The fiber of `X` is represented by some Galois object `A`:
+  Evaluation at some `a` in the fiber of `A` induces a bijection `A ⟶ X` to `F.obj X`.
+
+## References
+
+* H. W. Lenstra. Galois theory for schemes
```
Comment (jcommelin) [inline Mathlib/CategoryTheory/Galois/Decomposition.lean:29]:
```suggestion
* [lenstraGSchemes]: H. W. Lenstra. Galois theory for schemes
```

**Review by jcommelin — APPROVED** [review]
Thanks, and sorry for the delay in getting back to this.
1 minor comment.

bors d+

**Comment by chrisflav** [comment]
Thanks for the review.

**Pull request closed without merging**